tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc

```diff
@@ -451,26 +517,33 @@ LogicalResult SymbolicDimMgr::save() {
           (llvm::Twine("S") + llvm::Twine(numNonConstDims++)).str();
     }
   }
-  for (SymbolicDimOp op : usedSymbolicOps)
-    op.setName(nameMapping[op.getName().str()]);
+  std::unordered_map<std::string, SymbolicDimOp> name2Symbol;
+  for (SymbolicDimOp op : usedSymbolicOps) {
+    auto name = op.getName().str();
+    op.setName(nameMapping[name]);
+    name2Symbol[name] = op;
+  }
 
   // replace the name of a symbolic dim op to its new name.
-  if (failed(walkRankedTensorValue(m_, [&](Value value, RankedTensorType ty,
-                                           ArrayAttr attrs) {
-        SmallVector<Attribute> newAttrs;
-        for (Attribute attr : attrs) {
-          auto sym = m_.lookupSymbol<SymbolicDimOp>(
-              nameMapping[attr.cast<FlatSymbolRefAttr>().getValue().str()]);
-          assert(sym);
-          FlatSymbolRefAttr newAttr = FlatSymbolRefAttr::get(sym);
-          newAttrs.push_back(newAttr);
-        }
-        auto symbolicShapeAttr = ArrayAttr::get(value.getContext(), newAttrs);
-        auto newTy = RankedTensorType::get(ty.getShape(), ty.getElementType(),
-                                           symbolicShapeAttr);
-        value.setType(newTy);
-        return success();
-      }))) {
+  // update attributes attached to operations.
+  m_.walk([&](Operation* op) {
+    auto attrs =
+        op->getAttrOfType<ArrayAttr>(SymbolicDimOp::getSymbolicDimAttrName());
+    if (!attrs) return;
+    auto symbolicShapeAttr = updateAttrs(
+        attrs, [&](StringRef name) { return name2Symbol[name.str()]; });
+    op->setAttr(SymbolicDimOp::getSymbolicDimAttrName(), symbolicShapeAttr);
+  });
+  // update attributes attached to values.
+  if (failed(walkRankedTensorValue(
+          m_, [&](Value value, RankedTensorType ty, ArrayAttr attrs) {
+            auto symbolicShapeAttr = updateAttrs(
+                attrs, [&](StringRef name) { return name2Symbol[name.str()]; });
```

Comment:
nit: how about use `StringRef` as key to `name2Symbo`, it seems safe and no need to get a `std::string` instance from name each time.

Reply:
`StringRef` is not safe since it more like a pointer while the underline string buffer could be deleted before the map is deleted (e.g. some symbolic dim op is deleted or renamed). Thus we generally should not use StringRef as key if we can not guarantee it's safe.